### PR TITLE
Add RTLIL Backend.

### PR DESCRIPTION
**Pull request by nxmq**
### Contributor Checklist

- [ ] Did you add Scaladoc to every public function/method?
- [x] Did you update the FIRRTL spec to include every new feature/behavior?
- [x] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [x] Did you state the API impact?
- [x] Did you specify the code generation impact?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

- backend code generation
- new feature/API

#### API Impact

This PR adds the names "experimental-rtlil" and "rtlil" to the list of valid emitters, and also adds the annotations required to utilize the emitter. These are the only changes to the public-facing API.

#### Backend Code Generation Impact

This PR introduces an entirely new code generation backend, targeting the textual representation of the RTLIL IR format used by the Yosys open synthesis tool. As such, it should have minimal effects on the existing Verilog backend, as the only change made which could affect that backend was making verilog-specific memory lowering an optional dependency of the CombineCats pass, instead of a required one.

#### Desired Merge Strategy

Rebase please.

#### Release Notes
- Adds code generation backend targeting RTLIL IR for use with the Yosys open synthesis tool.

### Reviewer Checklist (only modified by reviewer)
- [x] Did you add the appropriate labels?
- [x] Did you mark the proper milestone (1.2.x, 1.3.0, 1.4.0) ?
- [x] Did you review?
- [x] Did you check whether all relevant Contributor checkboxes have been checked?
- [x] Did you mark as `Please Merge`?

### CODE REVIEW

File: src/main/scala/firrtl/Emitter.scala
Comment (ekiwi) [inline src/main/scala/firrtl/Emitter.scala:153]: Is your backend capable of emitting one file per module? If not, you shouldn't add it here.
The `smt2` backend for example always produces a single file which is why it is only supported by the `-E`  option.

File: src/main/scala/firrtl/Emitter.scala
Comment (ekiwi) [inline src/main/scala/firrtl/Emitter.scala:64]: For now, only `experimental-rtlil`. The SMT backend went through a year of updates until we decided it was time to allow the use without the `experimental-` prefix. Essentially by forcing users to acknowledge that this is experimental we reserve the right to radically change or remove the backend in the future.

File: src/main/scala/firrtl/stage/FirrtlAnnotations.scala
Comment (ekiwi) [inline src/main/scala/firrtl/stage/FirrtlAnnotations.scala:196]: I don't think this is needed here. Afaik this is only used by the old `CompilerAnnotation`.

File: src/main/scala/firrtl/transforms/CombineCats.scala
Comment (ekiwi) [inline src/main/scala/firrtl/transforms/CombineCats.scala:69]: This change is fine, shouldn't mess anything up for the Verilog emission :+1:

File: src/main/scala/firrtl/Emitter.scala
Comment (ekiwi) [reply src/main/scala/firrtl/Emitter.scala:153]: I just saw that you support `EmitAllModulesAnnotation` in your emitter code. You probably just forgot to add it here.

File: src/main/scala/firrtl/backends/experimental/rtlil/RtlilEmitter.scala
Comment (ekiwi) [inline src/main/scala/firrtl/backends/experimental/rtlil/RtlilEmitter.scala:21]: Since this is experimental, please make the `RtlilEmitter` object `private [firrtl]`.

File: src/main/scala/firrtl/backends/experimental/rtlil/RtlilEmitter.scala
Comment (ekiwi) [inline src/main/scala/firrtl/backends/experimental/rtlil/RtlilEmitter.scala:87]: Internal classes should always be `private`

File: src/main/scala/firrtl/backends/experimental/rtlil/RtlilEmitter.scala
Comment (ekiwi) [inline src/main/scala/firrtl/backends/experimental/rtlil/RtlilEmitter.scala:787]: all internal functions should be `private`

Comment (ekiwi) [reply src/main/scala/firrtl/backends/experimental/rtlil/RtlilEmitter.scala:787]: (and by internal I mean anything that isn't called from outside the class)

**Review by ekiwi — CHANGES_REQUESTED** [review]
This looks awesome. I added some small initial feedback.
Before this gets merged we need some tests. You can look at https://github.com/chipsalliance/firrtl/tree/master/src/test/scala/firrtl/backends/experimental/smt for inspiration.

A minimal smoke test could be to run your backend on all our regression firrtl files and then to just read the result with yosys to make sure it is at least valid RTLIL. Like: https://github.com/chipsalliance/firrtl/blob/master/src/test/scala/firrtl/backends/experimental/smt/end2end/SMTCompilationTest.scala

Thanks!

**Comment by nxmq** [comment]
> This looks awesome. I added some small initial feedback.
> Before this gets merged we need some tests. You can look at https://github.com/chipsalliance/firrtl/tree/master/src/test/scala/firrtl/backends/experimental/smt for inspiration.
> 
> A minimal smoke test could be to run your backend on all our regression firrtl files and then to just read the result with yosys to make sure it is at least valid RTLIL. Like: https://github.com/chipsalliance/firrtl/blob/master/src/test/scala/firrtl/backends/experimental/smt/end2end/SMTCompilationTest.scala
> 
> Thanks!

One test I did use throughout development was using yosys's `equiv_check` and related features to verify equivalence between the circuit as emitted by the rtlil backend and the circuit as emitted by the verilog backend. Unfortunately, this test takes about 5m to run on even a moderately complex circuit, so I feel hesitant about including something with such a long runtime into the tests for firrtl. Should I add an `equiv_check` based test, or a simple "does yosys consume this" test?

**Comment by ekiwi** [comment]
> One test I did use throughout development was using yosys's `equiv_check` and related features to verify equivalence between the circuit as emitted by the rtlil backend and the circuit as emitted by the verilog backend. Unfortunately, this test takes about 5m to run on even a moderately complex circuit, so I feel hesitant about including something with such a long runtime into the tests for firrtl. Should I add an `equiv_check` based test, or a simple "does yosys consume this" test?

I would go for "does yosys consume this" for all regression designs and then a `equiv_check` for a very small design.

**Comment by ekiwi** [comment]
To solve the CI issues, you just need to run:
```
sbt scalafmtAll
```

**Comment by ekiwi** [comment]
Not saying you need to do this to get the PR merged, but just as another idea on how we could test this backend:
You could use cxxrtl to generate an executable simulation with yosys and then you could write some executable unittests similar to what we do with Verilator to test the Verilog emission.
- [executable memory init test](https://github.com/chipsalliance/firrtl/blob/765af619af24fccd2dea748e95d70c9c6e8ea12b/src/test/scala/firrtlTests/MemoryInitSpec.scala#L359)
- [SimpleExecutionTest](https://github.com/chipsalliance/firrtl/blob/1f7faaa36816fc145a39bb7e736fd5ab739112c4/src/test/scala/firrtlTests/execution/SimpleExecutionTest.scala)

**Comment by ekiwi** [comment]
Looks like some of your code doesn't compile on scala 2.13. To debug that you can do `sbt ++2.13.4 compile`

**Comment by ekiwi** [comment]
@nxmq  We discussed this PR in our dev meeting and we are happy to merge it once the tests pass.

I assume that the reason the CI fails right now is that we might be using an older version of `yosys`. Could you add some output to stdout to debug what is going wrong with yosys?

**Comment by ekiwi** [comment]
Could you also add some of your formal equivalence check tests? Maybe marked as `ignore` for now? We would be interested in running them from time to time (but maybe not on every PR).

**Comment by ekiwi** [comment]
This PR might help us get a newer yosys version: https://github.com/chipsalliance/firrtl/pull/2365

File: src/main/scala/firrtl/Emitter.scala
Comment (ekiwi) [inline src/main/scala/firrtl/Emitter.scala:251]: please move these classes into the `firrtl.backends.experimental.rtlil` package similar to how we did it for the SMT backend : https://github.com/chipsalliance/firrtl/blob/a921e1230b389be87d993f3016cb46174b1ebfad/src/main/scala/firrtl/backends/experimental/smt/SMTEmitter.scala#L46

File: src/main/scala/firrtl/stage/FirrtlAnnotations.scala
Comment (ekiwi) [inline src/main/scala/firrtl/stage/FirrtlAnnotations.scala:234]: Please undo your changes to this file. This code is for the (soon to be) deprecated `-X` option. New emitters should only use the `-E` option.

File: src/main/scala/firrtl/transforms/CombineCats.scala
Comment (ekiwi) [inline src/main/scala/firrtl/transforms/CombineCats.scala:70]: :+1:

**Review by ekiwi — CHANGES_REQUESTED** [review]
Two small changes and then I think this new experimental backend is ready to :ship:

**Review by ekiwi — APPROVED** [review]
Awesome! Thank you very much for your patience and your hard work!

Looks like this change will make it into the upcoming release and it should be part of the snapshot in 1-2h. I would be happy to review any followup PRs if you see that something is missing or anything needs to be fixed.

**Pull request merged**
